### ISSUE TITLE
[Netty 5] Lambda-friendly future listeners

### ISSUE THREAD

**Issue by Bennett-Lynch** [netty/netty#9353]
Java 8, which introduced lambdas, allows for the following code:

```
channel.writeAndFlush(msg).addListener(new ChannelFutureListener() {
    @Override
    public void operationComplete(ChannelFuture future) {
        ...
    }
});
```

to become:

```
channel.writeAndFlush(msg).addListener((ChannelFutureListener) future -> {
    ...
});
```

However, this requires adding a type declaration to the lambda parameter, e.g., `(ChannelFutureListener)`.

Given the prominence of `ChannelFutureListener`, would it be worth adding a method signature to `ChannelFuture` that explicitly supports `ChannelFutureListener`?

Perhaps it would actually be worth renaming the current `addListener` in the `Future` interface to `addGenericListener` and then introducing type-explicit methods in extensions of `Future`.

E.g., the `ChannelFuture` interface would add:

```
ChannelFuture addListener(ChannelFutureListener listener);
```

Resulting in the slightly less verbose:

```
channel.writeAndFlush(msg).addListener(future -> {
    ...
});
```

**Comment by normanmaurer** [comment]
@Bennett-Lynch yeah... That said I think we may want to cleanup all of this a bit for Netty 5. See https://github.com/netty/netty/milestone/195

**Comment by Bennett-Lynch** [comment]
Thanks, Norman.

If I can sneak one more feature request in here, I'd also like to see `onSuccess()` and `onFailure()` convenience listeners added.  E.g.,

```
ctx.write(msg).addListener((ChannelFutureListener) future -> {
    if (future.isSuccess()) {
        System.out.println("success!");
    }
});
```

would become:

```
ctx.write(msg).onSuccess((ChannelFutureListener) future -> {
    System.out.println("success!");
});
```

And `ChannelFutureListener.CLOSE_ON_FAILURE` could be deprecated in favor of:

```
ctx.write(msg).onFailure(ChannelFutureListener.CLOSE);
```
